numba/tests/test_pythonapi_emptyargs.py

```diff
@@ -0,0 +1,81 @@
+from numba import njit
+from numba.core import types, cgutils, extending
+from numba.core.datamodel import registry, models
+import unittest
+
+
+class Dense():
+    def __init__(self, *, N: int):
+        self.N : int = N
+
+
+class DenseType(types.Type):
+    def __init__(self, *, N_type):
+        self.N_type = N_type
+        name : str = f"Dense<{N_type}>"
+        super().__init__(name)
+
+    # Type is mutable
+    mutable = True
+
+    @property
+    def key(self):
+        return self.N_type
+
+
+@registry.register_default(DenseType)
+class DenseModel(models.StructModel):
+    def __init__(self, dmm, fe_type: DenseType):
+        members = [
+            ("N", fe_type.N_type),
+        ]
+        super().__init__(dmm, fe_type, members)
+
+
+@extending.type_callable(Dense)
+def type_dense(context):
+    def typer(N):
+        return DenseType(N_type=N)
+    return typer
+
+
+@extending.lower_builtin(Dense, types.Any)
+def sparse_dense_constructor(context, builder, sig, args):
+    [N] = args
+    dense = cgutils.create_struct_proxy(sig.return_type)(context, builder)
+    dense.N = N
+    return dense._getvalue()
+
+
+@extending.box(DenseType)
+def box_dense(typ: DenseType, val, c):
+
+    dense = cgutils.create_struct_proxy(typ)(c.context, c.builder, value=val)
+
+    N_obj = c.pyapi.long_from_longlong(dense.N)
+    class_obj = c.pyapi.unserialize(c.pyapi.serialize_object(Dense))
+
+    kwds = c.pyapi.dict_pack({"N": N_obj}.items())
+    res = c.pyapi.call(class_obj, None, kwds)  # segfault
+
+    c.pyapi.decref(N_obj)
+    c.pyapi.decref(class_obj)
+    c.pyapi.decref(kwds)
+    return res
+
+
+@njit
+def foo(n: int):
+    d = Dense(N=n)
+    return d
+
+
+class PythonAPIEmptyArgs(unittest.TestCase):
+    def test_empty_args(self):
```

Comment:
I think this test should go in the `numba/tests/test_pythonapi.py` file.

Also, the test is too involved for what it is testing. I'm suggesting the following as an alternative:


```python
from numba import njit
from numba.core.types import none, intp
from numba.core.extending import intrinsic
import unittest
from numba.tests.support import captured_stdout


class PythonAPIEmptyArgs(unittest.TestCase):
    def test_empty_args(self):
        def callme(**kwargs):
            print("callme", kwargs)

        @intrinsic
        def py_call(tyctx):
            def codegen(context, builder, sig, args):
                pyapi = context.get_python_api(builder)
                gil = pyapi.gil_ensure()

                num = pyapi.long_from_long(context.get_constant(intp, 0xCAFE))
                kwds = pyapi.dict_pack({"key": num}.items())
                fn_print = pyapi.unserialize(pyapi.serialize_object(callme))
                # segfault: https://github.com/numba/numba/issues/5871
                res = pyapi.call(fn_print, None, kwds)

                pyapi.decref(res)
                pyapi.decref(fn_print)
                pyapi.decref(kwds)
                pyapi.decref(num)

                pyapi.gil_release(gil)
                return res

            return none(), codegen

        @njit
        def foo():
            py_call()

        with captured_stdout() as out:
            foo()
        d = {"key": 0xCAFE}
        expected = f"callme {d}\n"
        self.assertEqual(out.getvalue(), expected)


```